x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/join/Join.java

```diff
@@ -0,0 +1,144 @@
+        super(source, left, right);
+        this.config = config;
+    }
+
+    public Join(PlanStreamInput in) throws IOException {
+        super(in.readSource(), in.readLogicalPlanNode(), in.readLogicalPlanNode());
+        this.config = new JoinConfig(in);
+    }
+
+    public void writeTo(PlanStreamOutput out) throws IOException {
+        out.writeSource(source());
+        out.writeLogicalPlanNode(left());
+        out.writeLogicalPlanNode(right());
+        config.writeTo(out);
+    }
+
+    public JoinConfig config() {
+        return config;
+    }
+
+    @Override
+    protected NodeInfo<Join> info() {
+        return NodeInfo.create(this, Join::new, left(), right(), config);
+    }
+
+    @Override
+    public Join replaceChildren(List<LogicalPlan> newChildren) {
+        return new Join(source(), newChildren.get(0), newChildren.get(1), config);
+    }
+
+    public Join replaceChildren(LogicalPlan left, LogicalPlan right) {
+        return new Join(source(), left, right, config);
+    }
+
+    @Override
+    public List<Attribute> output() {
+        if (lazyOutput == null) {
+            lazyOutput = computeOutput();
+        }
+        return lazyOutput;
+    }
+
+    private List<Attribute> computeOutput() {
+        List<Attribute> right = makeReference(right().output());
+        return switch (config.type()) {
+            case LEFT -> // right side becomes nullable
+                NamedExpressions.mergeOutputAttributes(left().output(), makeNullable(right));
+            case RIGHT -> // left side becomes nullable
+                NamedExpressions.mergeOutputAttributes(makeNullable(left().output()), right);
+            case FULL -> // both sides become nullable
+                NamedExpressions.mergeOutputAttributes(makeNullable(left().output()), makeNullable(right));
+            default -> // neither side becomes nullable
+                NamedExpressions.mergeOutputAttributes(left().output(), right);
+        };
+    }
+
+    /**
+     * Make fields references, so we don't check if they exist in the index.
+     * We do this for fields that we know don't come from the index.
+     * TODO we should rework stats so we don't have to do this
+     */
+    public static List<Attribute> makeReference(List<Attribute> output) {
+        List<Attribute> out = new ArrayList<>(output.size());
+        for (Attribute a : output) {
+            out.add(new ReferenceAttribute(a.source(), a.name(), a.dataType(), a.qualifier(), a.nullable(), a.id(), a.synthetic()));
+        }
+        return out;
+    }
+
+    public static List<Attribute> makeNullable(List<Attribute> output) {
+        List<Attribute> out = new ArrayList<>(output.size());
+        for (Attribute a : output) {
+            out.add(a.withNullability(Nullability.TRUE));
+        }
+        return out;
+    }
+
+    @Override
+    public boolean expressionsResolved() {
+        return true;
```

Comment:
This doesn't take into account the validity of the join config. If I'm right, it's possible to define a broken `Join`, referring to non-existing fields in the join config, but have `resolved()` come out without any problems.